### PR TITLE
Longformer

### PR DESCRIPTION
**Pull request by ZihanWangKi**
# Description

> :memo: Please include a summary of the change. 
>  
> * Please also include relevant motivation and context.  
> * List any dependencies that are required for this change.  

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [x] A new research paper code implementation

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

`longformer_attention_test.py` and `longformer_encoder_test.py`.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: official/projects/longformer/longformer_encoder.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+# pylint: disable=g-classes-have-attributes
+
+from typing import Any, Callable, Optional, Union, List
+from absl import logging
+import tensorflow as tf
+
+from official.nlp.modeling import layers
+from official.projects.longformer.longformer_encoder_block import LongformerEncoderBlock
+
+def shape_list(tensor: tf.Tensor) -> List[int]:
```
Comment (CrickWu) [inline official/projects/longformer/longformer_encoder.py:28]: Making it private or can we simply use tf_utils.get_shape_list? https://github.com/tensorflow/models/blob/34a93745259ca40db718763a85ad38fc8ee433e5/official/modeling/tf_utils.py#L131

File: official/projects/longformer/longformer_encoder.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+      output_dropout: float = 0.1,
+      attention_dropout: float = 0.1,
+      initializer: _Initializer = tf.keras.initializers.TruncatedNormal(
+          stddev=0.02),
+      output_range: Optional[int] = None,
+      embedding_width: Optional[int] = None,
+      embedding_layer: Optional[tf.keras.layers.Layer] = None,
+      norm_first: bool = False,
+      **kwargs):
+    # Pops kwargs that are used in V1 implementation.
```
Comment (CrickWu) [inline official/projects/longformer/longformer_encoder.py:123]: Might not need these compatible lines if you are not using V1 BERT checkpoint initialization arguments.

File: official/projects/longformer/longformer_encoder.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+    norm_first: Whether to normalize inputs to attention and intermediate dense
+      layers. If set False, output of attention and intermediate dense layers is
+      normalized.
+  """
+
+  def __init__(
+      self,
+      vocab_size: int,
+      attention_window: Union[List[int], int] = 512,
+      global_attention_size: int = 0,
```
Comment (CrickWu) [inline official/projects/longformer/longformer_encoder.py:105]: Docstrings?

File: official/projects/longformer/longformer_encoder.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+  ):
+    """A helper function to pad tokens and mask to work with implementation of Longformer selfattention."""
+    # padding
+    attention_window = (
+        self._attention_window if isinstance(self._attention_window, int) else max(self._attention_window)
+    )
+
+    assert attention_window % 2 == 0, f"`attention_window` should be an even value. Given {attention_window}"
+
+    # input_shape = shape_list(input_ids) if input_ids is not None else shape_list(inputs_embeds)
```
Comment (CrickWu) [inline official/projects/longformer/longformer_encoder.py:364]: Delete? Or use the shape_list function below?

**Review by CrickWu — COMMENTED** [review]
Sending some initial comments.

File: official/projects/longformer/longformer_encoder.py (hunk 2 above)
Comment (ZihanWangKi) [reply official/projects/longformer/longformer_encoder.py:109]: I am using `intermediate_size` in longformer.py. If this is the usage in an old version, which file should I check to update to the newer version?

**Comment by ZihanWangKi** [comment]
Fixed comments & updated the code a bit to match with the master branch of tf models.
One thing not sure about is I am using an `any` encoder config, and in Line 230 of [sentence_prediction_with_checkpoint_convert.py](https://github.com/tensorflow/models/pull/10514/files#diff-f5ffd0a7c4bb552f0d632b5d580ce498b0c6966f4d3992e2c4969daa2060c66a) I am accessing the encoder config with `task_config.model.encoder.any`.

File: official/projects/longformer/longformer_encoder_test.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+import tensorflow as tf
+from absl.testing import parameterized
+from tensorflow.python.keras import keras_parameterized  # pylint: disable=g-direct-tensorflow-import
+from tensorflow.python.distribute import combinations
+
+from official.projects.longformer.longformer_encoder import LongformerEncoder
+
+@keras_parameterized.run_all_keras_modes
+class LongformerEncoderTest(keras_parameterized.TestCase):
+  @combinations.generate(combinations.combine(
```
Comment (saberkun) [inline official/projects/longformer/longformer_encoder_test.py:27]: add an empty line.

File: official/projects/longformer/sentence_prediction_with_checkpoint_convert.py [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+        encoder._transformer_layers[layer_num]._output_layer_norm.set_weights(
+            [allenai_model[f"encoder.layer.{layer_num}.output.LayerNorm.weight"],
+             allenai_model[f"encoder.layer.{layer_num}.output.LayerNorm.bias"]]
+        )
+    if not ckpt_dir_or_file:
+      return
+    if tf.io.gfile.isdir(ckpt_dir_or_file):
+      ckpt_dir_or_file = tf.train.latest_checkpoint(ckpt_dir_or_file)
+
+    pretrain2finetune_mapping = {
```
Comment (saberkun) [inline official/projects/longformer/sentence_prediction_with_checkpoint_convert.py:309]: tf.train.Checkpoint(encoder=encoder).write("path")

File: official/projects/longformer/sentence_prediction_with_checkpoint_convert.py [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+  hub_module_url: str = ''
+  metric_type: str = 'accuracy'
+  # Defines the concrete model config at instantiation time.
+  model: ModelConfig = ModelConfig()
+  train_data: cfg.DataConfig = cfg.DataConfig()
+  validation_data: cfg.DataConfig = cfg.DataConfig()
+
+
+@task_factory.register_task_cls(SentencePredictionConfig)
+class SentencePredictionTask(base_task.Task):
```
Comment (saberkun) [inline official/projects/longformer/sentence_prediction_with_checkpoint_convert.py:67]: Consider avoiding forking the task.
We can move initialize() to a standalone binary to save a tf checkpoint.
Or just add a write after initialization and keep this file for your local conversion use.

**Comment by saberkun** [comment]
If the comments are addressed, you could resolve the comments and ping us to review. Thanks! 
@CrickWu I feel Zihan updated. Please help review?

**Review by saberkun — APPROVED** [review]

**Pull request merged**